### ISSUE TITLE
Restoring Pod fails with Duplicate value for InitContainer Name

### ISSUE THREAD

**Issue by patst** [vmware-tanzu/velero#5193]
**What steps did you take and what happened:**
We use the OpenServiceMesh (OSM) Addon in an AKS cluster. (see https://openservicemesh.io/ )

In OSM enabled namespaces Pods get an InitContainer named `osm-init` and sidecar named `envoy` injected.

We have Pods with PersistentVolumes attached and using Restic for backup creation.
The backups are created with:
`velero backup create acc000-ns-def --include-namespaces acc000-example-box-def`

During restore we get a new persistent volume creating with dynamic provisioning. The volume is empty instead of containing the backed up resources.
If we use a namespace which is not OSM enabled the restore works.

The problem seems to be the initContainer and sidecar osm injects. In the logs we see these messages:
```
time="2022-08-10T07:40:32Z" level=error msg="Namespace acc000-example-box-def, resource restore error: error restoring pods/acc000-example-box-def/ps-velero-test-86bf7b5bd-2fst4: Pod \"ps-velero-test-86bf7b5bd-2fst4\" is invalid: [spec.volumes[2].name: Duplicate value: \"envoy-bootstrap-config-volume\", spec.containers[2].name: Duplicate value: \"envoy\", spec.initContainers[2].name: Duplicate value: \"osm-init\"]" logSource="pkg/controller/restore_controller.go:504" restore=velero/acc000-ns-def-20220810094016
```

**What did you expect to happen:**

Have the data restored
**The following information will help us better understand what's going on**:

Restore logs:

```
Name:         [1macc000-ns-def-20220810104615[0m
Namespace:    velero
Labels:       <none>
Annotations:  <none>

Phase:                       [31mPartiallyFailed[0m (run 'velero restore logs acc000-ns-def-20220810104615' for more information)
Total items to be restored:  51
Items restored:              51

Started:    2022-08-10 08:46:18 +0000 UTC
Completed:  2022-08-10 08:46:33 +0000 UTC

Warnings:
  Velero:     <none>
  Cluster:  could not restore, CustomResourceDefinition "certificaterequests.cert-manager.io" already exists. Warning: the in-cluster version is different than the backed-up version.
            could not restore, CustomResourceDefinition "certificates.cert-manager.io" already exists. Warning: the in-cluster version is different than the backed-up version.
            could not restore, CustomResourceDefinition "ingressbackends.policy.openservicemesh.io" already exists. Warning: the in-cluster version is different than the backed-up version.
            could not restore, CustomResourceDefinition "orders.acme.cert-manager.io" already exists. Warning: the in-cluster version is different than the backed-up version.
            could not restore, CustomResourceDefinition "podmonitors.monitoring.coreos.com" already exists. Warning: the in-cluster version is different than the backed-up version.
            could not restore, CustomResourceDefinition "postgresqlflexibles.database.gap.hdi.global" already exists. Warning: the in-cluster version is different than the backed-up version.
            could not restore, CustomResourceDefinition "scaledobjects.keda.sh" already exists. Warning: the in-cluster version is different than the backed-up version.
            could not restore, CustomResourceDefinition "secretstores.external-secrets.io" already exists. Warning: the in-cluster version is different than the backed-up version.
  Namespaces:
    acc000-example-box-def:  could not restore, ReplicaSet "acc000-def-example-8695cfdfcb" already exists. Warning: the in-cluster version is different than the backed-up version.
                             could not restore, Deployment "acc000-def-example" already exists. Warning: the in-cluster version is different than the backed-up version.
                             could not restore, Endpoints "acc000-def-example" already exists. Warning: the in-cluster version is different than the backed-up version.
                             could not restore, EndpointSlice "acc000-def-example-7c4nz" already exists. Warning: the in-cluster version is different than the backed-up version.
                             could not restore, Service "acc000-def-example" already exists. Warning: the in-cluster version is different than the backed-up version.

Errors:
  Velero:     <none>
  Cluster:    <none>
  Namespaces:
    acc000-example-box-def:  error restoring pods/acc000-example-box-def/ps-velero-test-86bf7b5bd-2fst4: Pod "ps-velero-test-86bf7b5bd-2fst4" is invalid: [spec.volumes[2].name: Duplicate value: "envoy-bootstrap-config-volume", spec.containers[2].name: Duplicate value: "envoy", spec.initContainers[2].name: Duplicate value: "osm-init"]

Backup:  acc000-ns-def

Namespaces:
  Included:  all namespaces found in the backup
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        nodes, events, events.events.k8s.io, backups.velero.io, restores.velero.io, resticrepositories.velero.io
  Cluster-scoped:  auto

Namespace mappings:  <none>

Label selector:  <none>

Restore PVs:  auto

Existing Resource Policy:   <none>

Preserve Service NodePorts:  auto
```

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Velero version (use `velero version`):  v1.9.0
- Velero features (use `velero client config get features`): features: <NOT SET>
- Kubernetes version (use `kubectl version`): v1.23.8
- Kubernetes installer & version: -
- Cloud provider or hardware configuration: Azure AKS
- OS (e.g. from `/etc/os-release`):

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

**Comment by ywk253100** [comment]
Seems the OSM injects the init containers during the pod creation while the restored pod also contains the same init containers.

Velero cannot handle such case at this moment.

**Comment by reasonerjt** [comment]
I think the webhook in OSM should be smart enough to skip injecting the init-container if the incoming request of the pod already has it?

**Comment by ywk253100** [comment]
One option is to implement one`RestoreItemAction` plugin to modify the pod specification and remove the init containers before the restoring, refer to the [doc](https://velero.io/docs/v1.9/overview-plugins/) for more info about the plugin

**Comment by patst** [comment]
@ywk253100 thanks for the advice.
We managed to implement a `RestoreItemAction` which edits the pod spec and removes the init container, sidecar and envoy configuration volume during restore. (https://github.com/patst/velero-plugin-osm-prune )

Meanwhile I opened a ticket in the OSM repository to suggest to fix it without an extra plugin https://github.com/openservicemesh/osm/issues/5008 .

I will close this issue.
